src/main/java/uk/ac/ebi/ampt2d/commons/accession/persistence/BasicSpringDataRepositoryDatabaseService.java

```diff
@@ -37,105 +42,189 @@
-        customMethodsRepository.insert(entitySet);
+    public AccessionWrapper<MODEL, String, ACCESSION> findAccession(ACCESSION accession)
+            throws AccessionDoesNotExistException, AccessionMergedException, AccessionDeprecatedException {
+        List<ACCESSION_ENTITY> entities = repository.findByAccession(accession);
+        assertAccessionIsActive(entities, accession);
+        return toAccessionWrapper(entities);
+    }
+
+    private void assertAccessionIsActive(List<ACCESSION_ENTITY> entities, ACCESSION accession)
+            throws AccessionDoesNotExistException, AccessionMergedException, AccessionDeprecatedException {
+        if (entities == null || entities.isEmpty()) {
+            assertAccessionMergedOrDeprecated(accession);
+            throw new AccessionDoesNotExistException(accession.toString());
+        }
+    }
+
+    private void assertAccessionMergedOrDeprecated(ACCESSION accession) throws AccessionDoesNotExistException,
+            AccessionMergedException, AccessionDeprecatedException {
+        IArchiveOperation<ACCESSION> operation = archiveService.getLastOperation(accession);
+        if (operation != null) {
+            switch (operation.getOperationType()) {
+                case MERGED_INTO:
+                    throw new AccessionMergedException(operation.getAccessionIdDestiny());
+                case DEPRECATED:
+                    throw new AccessionDeprecatedException();
+            }
+        }
+    }
+
+    private AccessionWrapper<MODEL, String, ACCESSION> toAccessionWrapper(List<ACCESSION_ENTITY> entities) {
+        final List<ModelWrapper<MODEL, String, ACCESSION>> models = entities.stream().map(this::toModelWrapper)
+                .collect(Collectors.toList());
+        return new AccessionWrapper<>(models);
     }
 
     @Override
-    public List<AccessionWrapper<MODEL, String, ACCESSION>> findAllAccessionMappingsByAccessions(
-            List<ACCESSION> accessions) {
-        List<AccessionWrapper<MODEL, String, ACCESSION>> result = new ArrayList<>();
+    public List<ModelWrapper<MODEL, String, ACCESSION>> findAllAccessions(List<ACCESSION> accessions) {
+        HashMap<ACCESSION, List<ACCESSION_ENTITY>> modelsByAccession = new HashMap<>();
         repository.findByAccessionIn(accessions).iterator().forEachRemaining(
-                entity -> result.add(createAccessionWrapperFromEntity(entity)));
-        return result;
+                entity -> {
+                    modelsByAccession.putIfAbsent(entity.getAccession(), new ArrayList<>());
+                    modelsByAccession.get(entity.getAccession()).add(entity);
+                });
+
+        return modelsByAccession.values().stream().map(this::filterOldVersions).map(this::toModelWrapper)
+                .collect(Collectors.toList());
+    }
+
+    private ACCESSION_ENTITY filterOldVersions(List<ACCESSION_ENTITY> accessionedElements) {
```

Comment:
Can the method name be getLatestVersion or lastVersion ?

Reply:
I would say `filterOldVersions` or `keepLastVersion`, but not a getter.